pyteal/compiler/compiler_test.py

```diff
@@ -967,7 +967,56 @@ def isEven(i: Expr) -> Expr:
 retsub
     """.strip()
     actual = compileTeal(program, Mode.Application, version=4, assembleConstants=False)
-    assert expected == actual
+    assert actual == expected
+
+
+def test_compile_subroutine_recursive_5():
+    @Subroutine(TealType.uint64)
+    def isEven(i: Expr) -> Expr:
+        return (
+            If(i == Int(0))
+            .Then(Int(1))
+            .ElseIf(i == Int(1))
+            .Then(Int(0))
+            .Else(isEven(i - Int(2)))
+        )
+
+    program = Return(isEven(Int(6)))
+
+    expected = """#pragma version 5
+int 6
+callsub sub0
+return
+sub0: // isEven
+store 0
+load 0
+int 0
+==
+bnz sub0_l5
+load 0
+int 1
+==
+bnz sub0_l4
+load 0
+int 2
+-
+load 0
+swap
+callsub sub0
+swap
+store 0
+sub0_l3:
+b sub0_l6
+sub0_l4:
+int 0
+b sub0_l3
+sub0_l5:
+int 1
+sub0_l6:
+retsub
+    """.strip()
+    actual = compileTeal(program, Mode.Application, version=5, assembleConstants=False)
+    assert actual == expected
 
 
```

Comment:
```suggestion


def test_compile_subroutine_recursive_cover_5():
    @Subroutine(TealType.uint64)
    def multiplyByAdding(a, b):
        return (
            If(a == Int(0))
            .Then(Return(Int(0)))
            .Else(
                Return(b + multiplyByAdding(a - Int(1), b))
            )
        )
    program = Return(multiplyByAdding(Int(3), Int(5)))

    expected = """#pragma version 5
int 3
int 5
callsub sub0
return
sub0: // multiplyByAdding
store 1
store 0
load 0
int 0
==
bnz sub0_l2
load 1
load 0
int 1
-
load 1
load 0
load 1
uncover 3
uncover 3
callsub sub0
cover 2
store 1
store 0
+
retsub
sub0_l2:
int 0
retsub
    """.strip()

    actual = compileTeal(program, Mode.Application, version=5, assembleConstants=False)
    assert actual == expected

```

If you don't mind, I think it might be great to add the example illustrated in PR, since it tests the `(un)cover` op-code optimization in subroutine recursion.

Reply:
Good idea, I just added this test and a few more